[SPARK-10149] [CORE] [WIP] Locality Level is always 'ANY' and performance is impacted heavily while reading from DFS file systems

wulei-bj-cn opened the pull request:
Corresponding JIRA issue: https://issues.apache.org/jira/browse/SPARK-10149

Starting from 1.4.1, locality level is always 'ANY' in HadoopRDD while reading from DFS file systems, which should have been 'NODE_LOCAL' for sure. This always being 'ANY' introduces lots of unnecessary network I/O where hadoop file blocks are transferred among nodes, which is a huge impact on overall performance. 

The main cause of this is : 
Function localHostName() in Utils.scala is trying to fetch the hostname for each of the hosts, yet when "SPARK_LOCAL_HOSTNAME" is not set, i.e. customHostname is null, this function will try to fetch the IP addresses for the hosts. That's because localIpAddress.getHostAddress is called, which will fetch the IP addresses in case customHostname is null. However, the returned IP addresses (1.2.3.4) will not match the hostnames (host1) that are fetched from DFS file systems. Hence locality level will always be 'ANY' and lots of network I/O is introduced when input files are read from DFS file systems. 

And solution for this issue:
Therefore, to make function return real hostnames when "SPARK_LOCAL_HOSTNAME" is not set, localIpAddress.getHostAddress is replaced by localIpAddress.getHostName, which will return a real hostname.

Testing of this patch:
Locality level returns to normal, i.e. 'NODE_LOCAL' , and network I/O that is introduced when locality level is 'ANY' is totally eliminated. This patch is tested on both standalone and YARN modes.

wulei-bj-cn committed:
Function localHostName() is trying to fetch the hostname for each of the hosts, yet when "SPARK_LOCAL_HOSTNAME" is not set, i.e. customHostname is null, this function will try to fetch the IP addresses for the hosts. That's because localIpAddress.getHostAddress is called, which will fetch the IP addresses in case customHostname is null. However, the returned IP addresses (1.2.3.4) will not match the hostnames (host1) that are fetched from DFS file systems. Hence locality level will always be 'ANY' and lots of network I/O is introduced when input files are read from DFS file systems. Therefore, to make function return real hostnames when "SPARK_LOCAL_HOSTNAME" is not set, localIpAddress.getHostAddress is replaced by localIpAddress.getHostName, which will return a real hostname.

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented:
Hm, no I think we made this return IP addresses on purpose to avoid other related problems. If you're saying it returns the 'wrong' IP address from among the possibilities, then this is what SPARK_LOCAL_HOSTNAME is for. see https://github.com/apache/spark/commit/9d117cee0be2c73a25702d98f78211055d50babe

wulei-bj-cn committed:
To make this patch less impact existing codes, I made a little adjustment. I still think that setting 'SPARK_LOCAL_HOSTNAME' to some specific value would be troublesome, especially when we need to deploy spark on hundreds or even thousands of nodes. Furthermore, hostname might change over time, keep updating 'SPARK_LOCAL_HOSTNAME' would be not practical in production environments. So a new boolean Spark env option 'SPARK_USE_HOSTNAME' is introduced to give end user an option to fetch local machines' default hostnames automatically. This will save lots of efforts in terms of deployment as well as maintenance of large scale Spark clusters.

wulei-bj-cn commented:
@srowen Thanks for your quick response! To make my patch less impact existing codes, I made a little adjustment. Would you plz help take a look again ? Thanks ! About Spark env option 'SPARK_LOCAL_HOSTNAME', I still insist that setting this env parameter to some specific value would be troublesome, especially when we deploy hundreds or even thousands of nodes on a large scale Spark cluster. Furthermore, hostnames might change over time, and keeping changing this 'SPARK_LOCAL_HOSTNAME' parameter would be not practical in production environments. That's why I'd like to introduce a new boolean Spark env option 'SPARK_USE_HOSTNAME' to make it easier for those people who want Spark to detect hostnames automatically. This option defaults to false, so it will not impact current codes.

wulei-bj-cn committed:
To make the code in this patch more Spark style.

srowen commented:
Why would the host name that happens to be returned be any more guaranteed to be the one you want? I don't think it needs yet another flag, as it just explodes the permutations of configuration.

This sounds like a problem to be solved elsewhere. You are merely concerned with identifying when two host identifiers are the same host even when they're two different identifiers. But is it even true that the several IPs or hostnames of a host are 'equivalent' in this sense? they could be different NICs.

wulei-bj-cn committed:
Make the patch be more Spark style.

wulei-bj-cn commented:
Dear Owen, thanks for checking my updates. I'm not saying this locality level being ANY all the time issue is caused by your code. Actually, it lies in code in org.apache.spark.scheduler.TaskSetManager:

// Check for node-local tasks
if (TaskLocality.isAllowed(locality, TaskLocality.NODE_LOCAL)) {
for (index <- speculatableTasks if canRunOnHost(index)) {
val locations = tasks(index).preferredLocations.map(_.host)
if (locations.contains(host))
{ speculatableTasks -= index return Some((index, TaskLocality.NODE_LOCAL)) }
}
}

The variable "locations" is hostnames of HDFS splits, which is from InetAddress.getHostName.
The variable "host" is IP address of an executor, which is from InetAddress.getHostAddress.

And this "host" variable's value is read from 
org.apache.spark.deploy.worker.WorkerArguments
where var host = Utils.localHostName()

Therefore, it leads to Utils.scala. I'm not saying we have to update Utils.scala to make things work, maybe we could update codes somewhere else to make this ANY go away too. Yet I just thought maybe updating a little bit code within Utils.scala would be kind of an easier way to do that. Or probably I'm wrong :)

Your solution of giving end user an option of "SPARK_LOCAL_HOSTNAME" works fine, given the tests I did with/without it on a Spark cluster of 4 nodes. No offense, but this setting is not typical in popular distributed computing systems. I mean, when it comes to deployment and maintenance, the configuration files (in our case, files under $SPARK_HOME/conf) should all be the same on all cluster nodes. However, this "SPARK_LOCAL_HOSTNAME" definitely will introduce differences on different nodes. And that's why I'd like to introduce a new setting "SPARK_USE_HOSTNAME", whose value could be the same on all cluster nodes, i.e. either "true" or "false".
 
About the multiple NICs you mentioned, I think it is a concern that OS should care about instead of our Spark.

srowen commented:
(I'm Sean) As I say, the problem with changing `Utils` is that you reintroduce problems elsewhere in the code. See the pull request I mentioned. A flag wouldn't help, right? if you change the behavior, it still causes a problem.

The root problem is that the worker is not returning the IP or host name that you think it should. At some level, you have to choose this if you need to choose this, so I don't see why `SPARK_LOCAL_HOSTNAME` is better or worse. I also don't understand why the host name is the one you want -- it may happen to be in your case, but this doesn't get at the more fundamental issue, that it need not be.

Aren't you really just trying to better identify when two different identifiers for a host are the same? that's not the way to solve this then.

wulei-bj-cn commented:
Basically, yes. Indeed this locality level being "ANY" is directly caused from org.apache.spark.scheduler.TaskSetManager:

// Check for node-local tasks
if (TaskLocality.isAllowed(locality, TaskLocality.NODE_LOCAL)) {
for (index <- speculatableTasks if canRunOnHost(index)) {
val locations = tasks(index).preferredLocations.map(_.host)
if (locations.contains(host))
{ speculatableTasks -= index return Some((index, TaskLocality.NODE_LOCAL)) }
}
}

As pointed out earlier, since variable "locations" contains java.lang.String values like "hostname1", "hostname2", etc and variable "host" is containing java.lang.String values like "1.2.3.4", "10.20.30.40", etc. So they will NEVER match, and that's why the task scheduler figures: "Ah ha, there is no task with preferred locality as NODE_LOCAL, and I'll just have to give it an ANY". 

Therefore, as far as I figured, we either change the "host" variable to contain values like hostnames or change the "locations" variable to contain IP addresses. The overall target is just to make them match when they're compared to each other. Am I correct ? 

I'd really hope you could help guide me in the right direction if I'm wrong :) Thanks so much !

srowen commented:
Agree, it seems like these need to be canonicalized to IP addresses. That feels like a right-er change, but I don't know enough to be sure. It's worth trying a change that would resolve the host (or IP) and then translate it into an IP always.

It still doesn't deal with the problem of reporting different interfaces or NICs but that's a slightly different higher-level problem, where you really would need something like `SPARK_LOCAL_HOSTNAME` since the workers can't otherwise know which interface to advertise on.

wulei-bj-cn commented:
Thanks Sean for your advice. As you suggested, I tried to translate hostnames to IP addresses in org.apache.spark.scheduler.TaskSetManager, and it turned out satisfying ('ANY' is eliminated and we saw 'NODE_LOCAL' again) after an initial testing. I'll do more verifications to make sure my patch works totally fine and won't introduce new bugs. I'll put the patch forward for your review after the verifications.

wulei-bj-cn committed:
As Sean suggested, giving up on modifying Utils.scala, try to resolve host names to IP addresses
when hostnames are not specified explicitly by user using SPARK_LOCAL_HOSTNAME.
TaskSetManager.scala is modified to support this hostname/IP address translation.

wulei-bj-cn commented:
Hi Sean,
As you suggested, I gave up modifying Utils.scala, and tried to resolve unspecified host names to IP addresses in org.apache.spark.scheduler.TaskSetManager. With this patch, I tried both scenarios with/without specifying SPARK_LOCAL_HOSTNAME manually, the testing results turned out to be satisfying, i.e. locality level 'ANY' from HadoopRDD is eliminated, and we got 'NODE_LOCAL' always where it's supposed to be. Would you plz help review my patch and see if it's suitable to be merged ? Thanks !

srowen commented:
@cmccabe or @kayousterhout or maybe @CodingCat could be a useful reviewer here.

srowen commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -190,11 +197,15 @@ private[spark] class TaskSetManager(
 
     for (loc <- tasks(index).preferredLocations) {
+
+      val locHost: String = if( sparkLocalHostname != None ) loc.host
```
Style issue: spaces outside parens, not inside. This could also be a `.getOrElse` construct maybe. Why does resolving it to an IP address depend on whether this host set `SPARK_LOCAL_HOSTNAME`?

wulei-bj-cn replied:
Yeah, right, I'll update accordingly to follow Spark/Scala styles. 

The reason why it depends on SPARK_LOCAL_HOSTNAME is : 
a) If SPARK_LOCAL_HOSTNAME is not set manually, 'Spark side', i.e. the Spark task scheduler,  will use IP address by default (as controlled from Utils.scala), but 'Hadoop side', i.e. HadoopRDD will still use host names, so we need to resolve HadoopRDD's host names to IP addresses, so that they could match with the 'Spark side'.
b) If SPARK_LOCAL_HOSTNAME is set manually, then the 'Spark side' will use host names instead(still, as controlled within Utils.scala), and as stated, the 'Hadoop side' always use host names, we wanna make sure we won't convert the 'Hadoop side' host names to IP addresses in this case, because they won't match with 'Spark side' host names otherwise.

That's why we need to first check if SPARK_LOCAL_HOSTNAME is set manually. The necessity of resolving host names (of HadoopRDDs) to IP addresses depends on this setting.

wulei-bj-cn committed:
Updated to align with Spark/Scala style check.

CodingCat commented:
Funny conversation, `"I'm Sean"`....

can we simply add `"SPARK_LOCAL_HOSTNAME = `hostname`"` to the spark-env.sh?

CodingCat commented:
don't know how to type it correctly 
![image](https://cloud.githubusercontent.com/assets/678008/9704543/59335564-5479-11e5-8581-2d262e7f4c0c.png)

CodingCat commented on core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala:
```diff
@@ -322,7 +333,12 @@ private[spark] class TaskSetManager(
       if (TaskLocality.isAllowed(locality, TaskLocality.NODE_LOCAL)) {
         for (index <- speculatableTasks if canRunOnHost(index)) {
-          val locations = tasks(index).preferredLocations.map(_.host)
+          val locations = if ( sparkLocalHostname != None ) {
```
@srowen, shall we have the spaces here?

wulei-bj-cn replied:
This patch has passed the Spark/Scala style check by running dev/run-tests.

CodingCat replied:
there are cases which we didn't cover in the scala style rules but are actually imposed "manually"....

e.g. even you remove the spaces around `=`, it still passes the scala style checks.....

CodingCat commented:
I mean to overcome the `large-scale configuration` problem mentioned by @wulei-bj-cn

wulei-bj-cn commented:
@CodingCat It's quite smart to set SPARK_LOCAL_HOSTNAME=`hostname` to save the efforts and energy I mentioned in earlier comments when it comes to deployment and maintenance. Yet, there are still chances where people/users forgot or just omitted this setting for some reason, since people tend to configure as less as possible. So we should still provide proper locality levels even when they have no idea about the existence of this setting, to make Spark more user-friendly. Am I correct or just over-reacting ? :)

CodingCat commented:
hmmm....I think the problem in this patch is that it involves too much network-relevant information in the scheduling modules. The ideal case shall be that the other module provides the HOSTNAME/HOST_IP to the scheduling module which always assumes this input to be correct....


can we move this changes to TaskLocation to make  the patch cleaner?

wulei-bj-cn committed:
As @CodingCat suggested, move the fix into TaskLocation.scala to make the path cleaner.

wulei-bj-cn committed:
Clear all the changes made in TaskSetManager.scala.

wulei-bj-cn commented:
@CodingCat Thanks so much for your suggestion! It does sound fair/smart to make the patch as cleaner as possible. As you suggested, I moved the fixes into TaskLocation.scala, and removed all the changes I made in TaskSetManager.scala, would you plz help take a look at the changes ? See if you've got more comments on it. 

BTW, the patch passed Spark/Scala style check, and I run a build with this patch on a 4-node cluster, results are good and expected. That is, no matter SPARK_LOCAL_HOSTNAME is set or not, we always get locality level as 'NODE_LOCAL' for HadoopRDDs.

CodingCat commented on core/src/main/scala/org/apache/spark/scheduler/TaskLocation.scala:
```diff
@@ -35,16 +37,26 @@ case class ExecutorCacheTaskLocation(override val host: String, executorId: Stri
 
 /**
  * A location on a host.
+ * The host names will be resolved to IP address in case SPARK_LOCAL_HOSTNAME is unset
```
You may want to refer "SPARK-10149" in the comments so that others can have more clues to understand the changes here....

CodingCat commented on core/src/main/scala/org/apache/spark/scheduler/TaskLocation.scala:
```diff
@@ -35,16 +37,26 @@ case class ExecutorCacheTaskLocation(override val host: String, executorId: Stri
  */
 private [spark] case class HDFSCacheTaskLocation(override val host: String) extends TaskLocation {
-  override def toString: String = TaskLocation.inMemoryLocationTag + host
+  override def toString: String = if ( sys.env.get("SPARK_LOCAL_HOSTNAME") == None ) {
```
would you like to move 'sys.env.get("SPARK_LOCAL_HOSTNAME")' to `object TaskLocation`?

CodingCat commented:
ok, this patch is looking good to me (after the changes I mentioned above)

time for others to judge @kayousterhout @srowen @andrewor14 @JoshRosen

kayousterhout commented:
I have all of the same reservations that @srowen mentioned -- I know we've had many issues in the past getting all of the addresses to match up in the way that they should, and I'm concerned that this change will introduce other issues with getting the hostnames to line up.  Happy to defer to @andrewor14 / @JoshRosen if either of you is convinced that this change lines up with all of the other ways that we use the host name / address elsewhere in the code.

If others decide this change is correct, one issue with this particular implementation is that it's very unclear to a naive reader of the code why the presence or absence of the SPARK_LOCAL_HOSTNAME variable should change how we interpret an address. I think one of the intermediate versions was more clear in this regard.

As an aside, one change that would help with this overall issue is if we more clearly named things as either host name or host IP address (the current variable names don't make this clear).

wulei-bj-cn committed:
Implementing @CodingCat's comments/suggestions in TaskLocation.scala.

CodingCat commented on core/src/main/scala/org/apache/spark/scheduler/TaskLocation.scala:
```diff
@@ -53,6 +67,10 @@ private[spark] object TaskLocation {
 
+  // Check if SPARK_LOCAL_HOSTNAME is set. If not set, host names for task locations will be
+  // resolved to IP addresses to fix the issue mentioned in SPARK-10149.
+  def sparkLocalHostname: Option[String] = sys.env.get("SPARK_LOCAL_HOSTNAME")
```
I mean, do we need to fetch the SPARK_LOCAL_HOSTNAME every time we reference it?

just make it as a `val` is fine, I guess

wulei-bj-cn commented:
@CodingCat Good point! Have just updated it to 'val'. See if there are any further comments/suggestions ?

wulei-bj-cn committed:
As @CodingCat suggested, make sparkLocalHostname a val instead of a def.

CodingCat commented:
@wulei-bj-cn , that's all in my side, thanks for your work

As @kayousterhout and @srowen  pointed out, we have changed time and time again on fetching the hostname/ip address of the server, which makes the correctness of this very fragile...as a result, this patch may break the assumption of other modules who need hostname/ip address (TaskLocation.scala might be fine, not sure.....)

Let's wait for the comments from others who are more confident about the changes we made around this.....

potto007 commented:
Can I suggest that, at the very least, SPARK_LOCAL_HOSTNAME become documented? I spent several hours trying to find why my latest Spark cluster was only choosing a locality of ANY. The existence of this PR was the only hint to me of the fix.

I'll take a stab at submitting a PR for the docs.

wulei-bj-cn commented:
Dear all,
I know it's been more than a month since I submitted this PR, and probably it's been rejected sliently. After a recent test with Spark + Tachyon, I found that set SPARK_LOCAL_HOSTNAME to 'hostname' will lead to locality level always being 'ANY' when Spark is reading file blocks from within Tachyon workers. Yet my patch could perfectly fix that locality level being 'ANY' issue when Spark is integrated with Tachyon. I'm not saying the community has to accept/merge my patch. I just hope that someone could take a deeper look into this 'ANY' issue and fix it as soon as possible. Since this issue starts to catch people's attention since 1.4.1, and soon enough, Spark 1.6.0 will be released. I really hope this issue is gone as soon as possible.

wulei-bj-cn commented:
And furthermore, this always being 'ANY' issue in Spark will cause Tachyon to keep re-caching remote memory blocks via networks from remote Tachyon workers, which introduces way too much unnecessary network I/Os, and what is even worse, more memory space is wasted due to this re-caching, since more than one copy of the file blocks are cached in the ram-disks. More details for this Spark + Tachyon issue could be found : https://groups.google.com/forum/#!topic/tachyon-users/HJ9xQH2AJxE .

srowen commented on core/src/main/scala/org/apache/spark/scheduler/TaskLocation.scala:
```diff
@@ -35,16 +37,28 @@ case class ExecutorCacheTaskLocation(override val host: String, executorId: Stri
  */
 private [spark] case class HostTaskLocation(override val host: String) extends TaskLocation {
-  override def toString: String = host
+  override def toString: String = if ( TaskLocation.sparkLocalHostname == None ) {
```
I think this is important too, I just don't know enough to evaluate the change. BTW these should be something more like `host.getOrElse(InetAddress.getByName(host).getHostAddress)`

I'd like to summon people like ... @andrewor14 @sryza @cmccabe. I do suspect this will be a problem in practice for several use cases.

wulei-bj-cn commented:
Dear all,
 @andrewor14 @sryza @cmccabe would you plz help take a look at this issue ? Thanks so much !

andrewor14 commented:
@wulei-bj-cn I'm not sure if I understand how your patch fixes the problem. What format does `HadoopRDD` expect? I looked at its `getPreferredLocations` method and it looks like the format it expects is hostname. Where is the `Utils.localHostName` that you're talking about?

steveloughran commented on core/src/main/scala/org/apache/spark/scheduler/TaskLocation.scala:
```diff
@@ -35,16 +37,28 @@ case class ExecutorCacheTaskLocation(override val host: String, executorId: Stri
 private [spark] case class HDFSCacheTaskLocation(override val host: String) extends TaskLocation {
-  override def toString: String = TaskLocation.inMemoryLocationTag + host
+  override def toString: String = if ( TaskLocation.sparkLocalHostname == None ) {
+    TaskLocation.inMemoryLocationTag + InetAddress.getByName(host).getHostAddress
```
better handle `InetAddress.getByName()` failure conditions here. It is entirely possible for a host to not know its own name if configured badly enough.

steveloughran commented:
Can I also suggest that the code is "IPv6 ready". Facebook have switched to IPv6 in their DCs, and [HADOOP-11890](https://issues.apache.org/jira/browse/HADOOP-11890) are working full time to migrate  ASF hadoop to supporting IPv6 alongside IPv4. The main thing here is probably to make sure that there aren't any assumptions about address type in storage or comparison of values. That won't magically mean that spark will work on IPv6 —only that the PR doesn't add one more bit of work for the future

rxin commented:
Thanks for the pull request. I'm going through a list of pull requests to cut them down since the sheer number is breaking some of the tooling we have. Due to lack of activity on this pull request, I'm going to push a commit to close it. Feel free to reopen it or create a new one.
